### ISSUE TITLE
Sky apps using mojo_shell on Android are crashing on launch

### ISSUE THREAD

**Issue by iansf** [domokit/mojo#318]
I would dive into this (and tried), but I no longer know of a working way to debug on Android.  mojodb fails with a hang when you run mojodb gdb_attach, and I don't see an option for gdb in mojo_shell.

The following command gives the following crash, which appears to have two different important failures ("GrGLInterface::validate() failed", and "Check failed: context_"):

```bash
$ mojo/tools/mojo_shell.py --android --sky sky/sdk/example/hello_world/lib/main.dart

Started http://127.0.0.1:31840 to host /Volumes/MojoHD/mojo/src/out/android_Debug.
Started http://127.0.0.1:9998 to host /Volumes/MojoHD/mojo/src.
--------- beginning of main
--------- beginning of system
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/authenticating_url_loader_interceptor.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/authenticating_url_loader_interceptor.mojo/.org.chromium.Chromium.GejcXo
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/tracing.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/tracing.mojo/.org.chromium.Chromium.Sv1NHq
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/debugger.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/debugger.mojo/.org.chromium.Chromium.CPIBXy
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/kiosk_wm.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/kiosk_wm.mojo/.org.chromium.Chromium.nM55rw
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/http_server.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/http_server.mojo/.org.chromium.Chromium.kRuY4T
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/view_manager.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/view_manager.mojo/.org.chromium.Chromium.OtYcSz
[ERROR:platform_viewport_android.cc(177)] Not implemented reached in virtual void native_viewport::PlatformViewportAndroid::SetBounds(const gfx::Rect&)
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/surfaces_service.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/surfaces_service.mojo/.org.chromium.Chromium.EghzIv
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/sky_viewer.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t
[INFO:network_fetcher.cc(87)] Caching mojo app http://127.0.0.1:31840/icu_data.mojo at /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a//127.0.0.1#3a31840/icu_data.mojo/.org.chromium.Chromium.xXtBcv
[0713/191327:INFO:GrGLInterface.cpp(208)] ../../third_party/skia/src/gpu/gl/GrGLInterface.cpp:208 GrGLInterface::validate() failed.

[0713/191327:FATAL:ganesh_context.cc(57)] Check failed: context_.
#00 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#01 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#02 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#03 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#04 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#05 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#06 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#07 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#08 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#09 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#10 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#11 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#12 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#13 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#14 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#15 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#16 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#17 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#18 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#19 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#20 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#21 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#22 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#23 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#24 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#25 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#26 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#27 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#28 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#29 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#30 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#31 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#32 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#33 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#34 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#35 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#36 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#37 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#38 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#39 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#40 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#41 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#42 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#43 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#44 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#45 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#46 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#47 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#48 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#49 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#50 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#51 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#52 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#53 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#54 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#55 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#56 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#57 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#58 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#59 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#60 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f
#61 0x9ba11b2f /data/data/org.chromium.mojo.shell/app_home/.mojo_url_response_disk_cache/http#3a/127.0.0.1#3a31840/sky_viewer.mojo/.org.chromium.Chromium.FXQ34t+0x006a6b2f

--------- beginning of crash
```

**Comment by iansf** [comment]
@qsr: mentioning you here in case you have any insight.

**Comment by cpu-chromium** [comment]
That is in skia. I suggest you try reverting the last, or the two last skia rolls (which I think is via deps) rebuild and try again.

If the problem goes away you want to ping reed@

**Comment by abarth** [comment]
Please don't revert the Skia rolls.  The problem is in the Mojo bindings to Skia.

**Comment by cdotstout** [comment]
Seem to be happening on linux desktop as well:

cstout@soldierjane:~/projects/mojo/src$ mojo/tools/mojo_shell.py
'mojo:kiosk_wm mojo:ganesh_app'
[0713/182731:INFO:GrGLInterface.cpp(208)]
../../third_party/skia/src/gpu/gl/GrGLInterface.cpp:208
GrGLInterface::validate() failed.

**Comment by abarth** [comment]
Yes.  That error means that Skia is rejecting the GrGLInterface that Mojo is giving to Skia.  The fix is to improve the GrGLInterface that Mojo builds so that it passes Skia's validation.

**Comment by tonygentilcore** [comment]
@iansf I'm becoming blocked on this, so I just wanted to check whether you're able to make progress w/ gdb using @cdotstout's desktop repro or if I or someone else should jump in to help.

/cc @ppiet @qsr @emembrives How far are we from gdb support in mojo devtools?

**Comment by tonygentilcore** [comment]
nm, fix in https://codereview.chromium.org/1236263005

**Comment by tonygentilcore** [comment]
So, that change seems to do the trick, but I'm confused by a few things:

- Why didn't our bots crash? Should we run sky tests in mojo shell on the bots?
- It looks like this file isn't rolled in from chromium any more because the fix landed there 3 months ago and I know we've rolled since then. Is that correct? If so, who is maintaining this code?
- Why do sky and mojo shell go down different code paths here?

**Comment by abarth** [comment]
> Why didn't our bots crash?

We don't have any testing of Skia inside Mojo.

> Should we run sky tests in mojo shell on the bots?

We used to run //sky/tests in mojo_shell, but we recently switched these over to a simpler testing harness.  The goal here is to make //sky/tests into normal Dart unit tests run by Dart's test_runner framework.

> It looks like this file isn't rolled in from chromium any more because the fix
> landed there 3 months ago and I know we've rolled since then. Is that correct?

That's correct.  We only pull base, build, and their dependencies from Chromium.  We don't pull anything higher in the dependency tree.

> If so, who is maintaining this code?

No one that I'm aware of.

> Why do sky and mojo shell go down different code paths here?

sky_shell doesn't use the GPU command buffer because there's only one stream of GL commands in a single process.  mojo_shell does use the GPU command buffer to (a) multiplex multiple streams of GL commands from each Mojo app and to (b) transmit GL commands from one process (i.e., Mojo app) to another.
